### PR TITLE
handle racing conditions when props are setted on exoplayer

### PR DESCRIPTION
**Pull request by CHaNGeTe**
As react native sets props one by one and not in batch, it seems it was producing some inconsistent scenarios.

Test this branch carefully, as some prop setting may need adjustment.
I have seen most setters check `if (player != null) {` but I may missed some

### CODE REVIEW

**Comment by CHaNGeTe** [comment]
Seems to solve issues like:
https://github.com/react-native-community/react-native-video/issues/1538

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
Comment (cobarx) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:345]: I would add a comment explaining that this is there to handle the race conditions.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
Comment (cobarx) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: Extra newline can be removed.

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
Comment (cobarx) [inline android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:1180]: This appears to be an unused variable.

**Review by cobarx — APPROVED** [review]
Didn't test this, but the code looks good reading through it.

Left you a few minor notes on stuff you might want to cleanup.

**Comment by gsunsnackv** [comment]
Awesome job @CHaNGeTe ! I confirm that particular pull request fix https://github.com/react-native-community/react-native-video/issues/1538. Code at https://github.com/gsunsnackv/react-native-player-stuck/tree/handle-props-racing-conditions-on-expplayer shows it works

I wonder if this will solve the black screen problem in general. Fingers crossed.

Please merge and release!

**Comment by CHaNGeTe** [comment]
Thank you so much @gsunsnackv 
Please have a little patience with the merge, as it is a huge change that affect all property setters.
I have been trying to be careful, but testing is super welcome.
I have seen this PR solves some problems on various scenarios.
It will be a great feedback if you can test, and mention this PR on all the black screen issues.
I am waiting for the opinion and feedback of other maintainers, but all test done in this branch is a positive step forward.

In the meantime you can add it to your app via `npm i -S 24i/react-native-video#handle-props-racing-conditions-on-expplayer`

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
Comment (CHaNGeTe) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:345]: commented

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
Comment (CHaNGeTe) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java]: deleted

File: android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java
Comment (CHaNGeTe) [reply android-exoplayer/src/main/java/com/brentvatne/exoplayer/ReactExoplayerView.java:1180]: It is being used there:
https://github.com/react-native-community/react-native-video/pull/1655/files/40c7371b3e8955ffae516dbda1a8783952c14381#diff-2b2f68a93f154d6095529ae77eeb90f1R395

in this line `this` is explicit because argument is named the same

**Comment by Balasnest** [comment]
`npm install 24i/react-native-video#feature/handle-props-racing-conditions-on-expplayer --save`

@CHaNGeTe I am unable to run and make a release build from this branch. 
@gsunsnackv Are you able to do it?

```
"react": "16.5.0",
"react-native": "0.57.2",
```
Getting this exception:

Screen Shot 2019-07-11 at 11 06 21 AM

```
Execution failed for task ':app:transformClassesWithDesugarForQaDebug'.
> com.android.build.api.transform.TransformException: java.lang.RuntimeException: java.lang.RuntimeException:
```

**Comment by gsunsnackv** [comment]
@Balasnest I did not try to release a build. I used the command `yarn git+https://github.com/24i/react-native-video#feature/handle-props-racing-conditions-on-expplayer` similar to what @CHaNGeTe suggested. That worked for me to run the app and no more black screen

**Comment by Balasnest** [comment]
> @Balasnest I did not try to release a build. I used the command `yarn git+https://github.com/24i/react-native-video#feature/handle-props-racing-conditions-on-expplayer` similar to what @CHaNGeTe suggested. That worked for me to run the app and no more black screen


@gsunsnackv It's working on your repo (`react-native: 0.59`). But I am getting exception error in my project (`react-native: 0.57`).

@CHaNGeTe Any idea why it is happening? Thanks

**Comment by Balasnest** [comment]
@CHaNGeTe @gsunsnackv Finally! It's working fine now. This pull request has a fix for this.

If anyone facing build issue, try this:
1. install lib from this branch
2. clean android gradle
3. run build or release apk.

**Comment by Balasnest** [comment]
@CHaNGeTe @jenshandersson @n1ru4l @ashnfb  Can we merge this PR as soon as possible? we are badly in need of our prod release. Please look into this. Thanks in advance.

**Review by ashnfb — APPROVED** [review]

**Comment by ashnfb** [comment]
@Balasnest merged into master

**Comment by CHaNGeTe** [comment]
@ashnfb this is unreleased, right?

**Comment by CHaNGeTe** [comment]
Released to 4.4.4

**Comment by salhotra** [comment]
Thanks for the PR guys but I can confirm #1649 & #1151 were not resolved by this change.

**Comment by IamMasterWayne** [comment]
"react-native": "0.61.5"
"react-native-video": "^5.0.2"

React Native Android Black Screen For Android

**Pull request merged**
